flink-libraries/flink-gelly/src/main/java/org/apache/flink/graph/library/AffinityPropagation.java

```diff
@@ -0,0 +1,535 @@
+		DataSet<Vertex<Long, APVertexValue>> APVerticesDS = input.getContext().fromCollection(APvertices);
+
+		ScatterGatherConfiguration parameters = new ScatterGatherConfiguration();
+		parameters.registerAggregator("convergedAggregator", new LongSumAggregator());
+
+		Graph<Long, APVertexValue, NullValue> APgraph
+			= Graph.fromDataSet(APVerticesDS, APEdgesDS, input.getContext());
+
+		return APgraph.getUndirected().runScatterGatherIteration(new APVertexUpdater(input.numberOfVertices() * 2),
+			new APMessenger(),this.maxIterations,parameters).getVertices().filter(new FilterFunction<Vertex<Long, APVertexValue>>() {
+			@Override
+			public boolean filter(Vertex<Long, APVertexValue> vertex) throws Exception {
+				return vertex.getId()%2 == 0;
+			}
+		}).map(new MapFunction<Vertex<Long, APVertexValue>, Tuple2<Long, Long>>() {
+			@Override
+			public Tuple2<Long, Long> map(Vertex<Long, APVertexValue> value) throws Exception {
+				Tuple2<Long, Long> returnValue = new Tuple2<>(value.getId()/10, value.getValue().getExemplar()/10);
+				return returnValue;
+			}
+		});
+
+	}
+
+	/**
+	* Foreach input point we have to create a pair of E,I vertices. Same structure is used for both vertex type, to
+	* diferenciate E and I vertices is used the id. Foreach input point we will create:
+	*
+	* - One E vertex with the id as the original input id * 10 + 1
+	* - One I vertex with the id as the original input id * 10
+	*
+	* This way even ids are from E type vertices and odd ids are from I vertices.
+	*
+	* It also calculates adds the weights to the I vertices. Notice that the S vertices are not created and the weights
+	* are added to the I vertices, simulating the S vertex.
+	*/
+
+	@SuppressWarnings("serial")
+	private static final class InitAPVertex implements EdgesFunction<Long, Double, Vertex<Long, APVertexValue>> {
+
+		@Override
+		public void iterateEdges(Iterable<Tuple2<Long, Edge<Long, Double>>> edges,
+								Collector<Vertex<Long, APVertexValue>> out) throws Exception {
+
+			Vertex<Long, APVertexValue> APvertexI = new Vertex<>();
+			Vertex<Long, APVertexValue> APvertexE = new Vertex<>();
+
+			Iterator<Tuple2<Long, Edge<Long, Double>>> itr = edges.iterator();
+			Tuple2<Long, Edge<Long, Double>> edge = itr.next();
+
+			APvertexE.setId(edge.f0 * 10 + 1);
+			APvertexE.setValue(new APVertexValue());
+
+			APvertexI.setId(edge.f0 * 10);
+			APvertexI.setValue(new APVertexValue());
+			APvertexI.getValue().getWeights().put(edge.f1.getSource() * 10 + 1, edge.f1.getValue());
+
+			APvertexE.getValue().getOldValues().put(edge.f1.getSource() * 10, 0.0);
+			APvertexI.getValue().getOldValues().put(edge.f1.getSource() * 10 + 1, 0.0);
+
+
+			while(itr.hasNext()){
+				edge = itr.next();
+				APvertexI.getValue().getWeights().put(edge.f1.getSource() * 10 + 1, edge.f1.getValue());
+
+				APvertexE.getValue().getOldValues().put(edge.f1.getSource() * 10, 0.0);
+				APvertexI.getValue().getOldValues().put(edge.f1.getSource() * 10 + 1, 0.0);
```

Comment:
So, for each vertex, you are creating two extra vertices, each of which basically carries two hashmaps in the order of neighbors, right? This is really of prohibitive cost for any graph with mid- or high-degree vertices. Is there any other way to implement this?